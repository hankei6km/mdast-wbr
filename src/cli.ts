import { Readable, Writable } from 'node:stream'
import { fromMarkdown } from 'mdast-util-from-markdown'
import { toMarkdown } from 'mdast-util-to-markdown'
import { frontmatter } from 'micromark-extension-frontmatter'
import {
  frontmatterFromMarkdown,
  frontmatterToMarkdown
} from 'mdast-util-frontmatter'
import { WbrOptions } from './lib/types.js'
import { wbr } from './lib/wbr.js'

type Opts = {
  stdin: Readable
  stdout: Writable
  stderr: Writable
} & Pick<WbrOptions, 'locales' | 'asStr'>

export const cli = async ({
  locales,
  asStr,
  stdin,
  stdout,
  stderr
}: Opts): Promise<number> => {
  try {
    let md = ''
    for await (const i of stdin) {
      md = md + i.toString('utf-8')
    }
    const tree = fromMarkdown(md, {
      extensions: [frontmatter(['yaml', 'toml'])],
      mdastExtensions: [frontmatterFromMarkdown(['yaml', 'toml'])]
    })
    stdout.write(
      toMarkdown(wbr(tree, { locales, asStr }), {
        extensions: [frontmatterToMarkdown(['yaml', 'toml'])]
      })
    )
  } catch (err: any) {
    stderr.write(err.toString())
    stderr.write('\n')
    return 1
  }
  return 0
}
