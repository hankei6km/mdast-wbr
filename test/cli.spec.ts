import { Readable, PassThrough } from 'stream'
import { cli } from '../src/cli.js'

describe('cli()', () => {
  // cli では for await..of で読んでいるので AsyncGenerator でも動く
  async function* mockStdin(md: string) {
    yield md
  }
  it('should print markdown word-breaked(html)', async () => {
    const stdin = mockStdin(
      '# 単語の区切り **ここは強調** 続きの文字列\nここはパラグラフ'
    ) as unknown as Readable
    const stdout = new PassThrough()
    const stderr = new PassThrough()
    let outData = ''
    stdout.on('data', (d) => (outData = outData + d))
    let errData = ''
    stderr.on('data', (d) => (errData = errData + d))
    expect(
      await cli({
        locales: 'ja',
        asStr: '',
        stdin,
        stdout,
        stderr
      })
    ).toEqual(0)
    expect(outData).toMatchSnapshot()
    expect(errData).toEqual('')
  })

  it('should print markdown word-breaked(frontmatter)', async () => {
    const stdin = mockStdin(
      '---\nabc: 1\n---\n\n# 単語の区切り **ここは強調** 続きの文字列\nここはパラグラフ'
    ) as unknown as Readable
    const stdout = new PassThrough()
    const stderr = new PassThrough()
    let outData = ''
    stdout.on('data', (d) => (outData = outData + d))
    let errData = ''
    stderr.on('data', (d) => (errData = errData + d))
    expect(
      await cli({
        locales: 'ja',
        asStr: '',
        stdin,
        stdout,
        stderr
      })
    ).toEqual(0)
    expect(outData).toMatchSnapshot()
    expect(errData).toEqual('')
  })

  it('should print markdown word-breaked(str)', async () => {
    const stdin = mockStdin(
      '# 単語の区切り **ここは強調** 続きの文字列\nここはパラグラフ'
    ) as unknown as Readable
    const stdout = new PassThrough()
    const stderr = new PassThrough()
    let outData = ''
    stdout.on('data', (d) => (outData = outData + d))
    let errData = ''
    stderr.on('data', (d) => (errData = errData + d))
    expect(
      await cli({
        locales: 'ja',
        asStr: '|',
        stdin,
        stdout,
        stderr
      })
    ).toEqual(0)
    expect(outData).toMatchSnapshot()
    expect(errData).toEqual('')
  })
})
