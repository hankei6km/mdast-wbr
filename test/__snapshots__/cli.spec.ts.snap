// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`cli() should print markdown word-breaked(frontmatter) 1`] = `
"---
abc: 1
---

# 単語<wbr>の<wbr>区切り<wbr> **ここは強調** <wbr>続き<wbr>の<wbr>文字<wbr>列

ここはパラグラフ
"
`;

exports[`cli() should print markdown word-breaked(html) 1`] = `
"# 単語<wbr>の<wbr>区切り<wbr> **ここは強調** <wbr>続き<wbr>の<wbr>文字<wbr>列

ここはパラグラフ
"
`;

exports[`cli() should print markdown word-breaked(str) 1`] = `
"# 単語|の|区切り| **ここは強調** |続き|の|文字|列

ここはパラグラフ
"
`;
